tools/auth/notion.go

```diff
@@ -0,0 +1,88 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"golang.org/x/oauth2"
+	"net/http"
+)
+
+var _ Provider = (*Notion)(nil)
+
+// NameNotion is the unique name of the Patreon provider.
+const NameNotion string = "notion"
+
+// Notion allows authentication via Notion OAuth2.
+type Notion struct {
+	*baseProvider
+}
+
+// NewNotionProvider creates new Notion provider instance with some defaults.
+func NewNotionProvider() *Notion {
+	return &Notion{&baseProvider{
+		ctx:        context.Background(),
+		authUrl:    "https://api.notion.com/v1/oauth/authorize",
+		tokenUrl:   "https://api.notion.com/v1/oauth/token",
+		userApiUrl: "https://api.notion.com/v1/users/me",
+	}}
+}
+
+// FetchAuthUser returns an AuthUser instance based on the Notion's User api.
+// API reference: https://developers.notion.com/reference/get-self
+func (p *Notion) FetchAuthUser(token *oauth2.Token) (*AuthUser, error) {
+	req, err := http.NewRequest("GET", p.userApiUrl, nil)
```

Comment:
Replace `NewRequest` with `NewRequestWithContext` and pass `p.ctx`. For example:

```go
req, err := http.NewRequestWithContext(p.ctx, "GET", p.userApiUrl, nil)
```